### PR TITLE
One-time migration task 20241108

### PR DESCRIPTION
**Pull request by gigxz**
## _Merging this PR_
- use the squash-merge strategy

## Description
Issue: https://github.com/open-path/Green-River/issues/6912

One-off migration task. This will need to be run before it's deployed, but figure it should go into into the codebase anyway, and get some review.

Sorry it's a bit of a long mess. See https://github.com/open-path/Green-River/issues/6912#issuecomment-2463110334 for more details.

## Type of change
Data migration

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [ ] I have updated the documentation (or not applicable)
- [ ] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

### CODE REVIEW

File: drivers/hmis_external_apis/lib/tasks/import.rake [hunk 1]
```diff
@@ -115,4 +115,220 @@ namespace :import do
...
+      25 => 'Shelter',
+      26 => 'Hotel',
+    }
+
+    data_source = GrdaWarehouse::DataSource.hmis.first
+    system_hud_user = Hmis::Hud::User.system_user(data_source_id: data_source.id)
+    project = Hmis::Hud::Project.hmis.find(args.project_id)
+
+    # Read Export file. Each row is a "Case" which represents a HoH Enrollment. It gives us the Entry Date.
+    export_file = HmisExternalApis::TcHmis::Importers::Loaders::XlsxFile.new(filename: "#{dir}/HCM-EXPORT.xlsx", sheet_number: 0, header_row_number: 1)
```
Comment (gigxz) [inline drivers/hmis_external_apis/lib/tasks/import.rake:173]: Reused this TcHmis XlsxFile class and it was helpful. Maybe we should move it to a generic utility class -- I'm not sure which `driver` that would go in? Or, I could just move it to another folder within `HmisExternalApis` (out of TcHmis) with the intention to use for any future custom imports that use xlsx.

File: drivers/hmis_external_apis/lib/tasks/import.rake [hunk 2]
```diff
@@ -115,4 +115,220 @@ namespace :import do
...
+      raise "Referral ID #{referral_id} linked to multiple households. unexpected." if referral_id_to_hh_id.key?(referral_id) && referral_id_to_hh_id[referral_id] != household_id
+
+      referral_id_to_hh_id[referral_id] = household_id
+
+      referral_id_to_household[referral_id] ||= []
+      referral_id_to_household[referral_id] << { mci_id: mci_id, relation_to_id: relation_to_id }
+    end
+
+    # Find Clients. No special care taken to situation where there are multiple clients with the same MCI ID, just pick one.
+    mci_id_to_client = HmisExternalApis::ExternalId.mci_ids.where(value: mci_ids).preload(:source).index_by(&:value)
```
Comment (ttoomey) [inline drivers/hmis_external_apis/lib/tasks/import.rake]: perhaps add an order so if there are duplicates, at least it's deterministic 
```suggestion
    mci_id_to_client = HmisExternalApis::ExternalId.mci_ids.where(value: mci_ids).order(:id).preload(:source).index_by(&:value)
```

File: drivers/hmis_external_apis/lib/tasks/import.rake [hunk 3]
```diff
@@ -115,4 +115,220 @@ namespace :import do
...
+    # Read Export file. Each row is a "Case" which represents a HoH Enrollment. It gives us the Entry Date.
+    export_file = HmisExternalApis::TcHmis::Importers::Loaders::XlsxFile.new(filename: "#{dir}/HCM-EXPORT.xlsx", sheet_number: 0, header_row_number: 1)
+
+    skipped_case_ids = []
+    referral_id_to_case_id = {}
+    referral_id_to_entry_date = {}
+    export_file.each do |row|
+      referral_id = row.field_value('ReferralID')
+      case_id = row.field_value('CASE_ID')
+      entry_date_time = row.field_value('OPEN_DT')
```
Comment (ttoomey) [inline drivers/hmis_external_apis/lib/tasks/import.rake:181]: should there be format validation or constraints on this date?

File: drivers/hmis_external_apis/lib/tasks/import.rake [hunk 4]
```diff
@@ -115,4 +115,220 @@ namespace :import do
...
+
+      household = referral_id_to_household[referral_id]
+      if !household
+        Rails.logger.info "Household not found for referral ID #{referral_id}. Skipping."
+        skipped_case_ids << case_id
+        next
+      end
+
+      household_id = Hmis::Hud::Base.generate_uuid
+      household.each do |member|
```
Comment (ttoomey) [inline drivers/hmis_external_apis/lib/tasks/import.rake:248]: should we check for duplicate household members? (one client in the same household multiple times)

File: drivers/hmis_external_apis/lib/tasks/import.rake [hunk 5]
```diff
@@ -115,4 +115,220 @@ namespace :import do
...
+      case_notes << Hmis::Hud::CustomCaseNote.new(
+        CustomCaseNoteID: Hmis::Hud::Base.generate_uuid,
+        EnrollmentID: enrollment.EnrollmentID,
+        PersonalID: enrollment.PersonalID,
+        data_source_id: enrollment.data_source_id,
+        content: note_content,
+        UserID: user_id,
+        DateCreated: date_created,
+        DateUpdated: date_updated,
+        information_date: information_date.to_date,
```
Comment (ttoomey) [inline drivers/hmis_external_apis/lib/tasks/import.rake:310]: should these all be '.to_date' ?

File: drivers/hmis_external_apis/lib/tasks/import.rake (hunk 1 above)
Comment (ttoomey) [reply drivers/hmis_external_apis/lib/tasks/import.rake:173]: 👍 to making it generic. Maybe HmisExternalApis::Importers::XlsxFile

**Review by ttoomey — APPROVED** [review]
Looks good!

File: drivers/hmis_external_apis/lib/tasks/import.rake (hunk 5 above)
Comment (gigxz) [reply drivers/hmis_external_apis/lib/tasks/import.rake:310]: No, only Information Date is a date. The timsetamps are datetimes (which are in the file). Double-checked that these are being set correctly

File: drivers/hmis_external_apis/lib/tasks/import.rake (hunk 4 above)
Comment (gigxz) [reply drivers/hmis_external_apis/lib/tasks/import.rake:248]: I dont think so, that's an acceptable data quality issue. Similarly I am not checking for multiple HoHs or no HoHs. Those are things they can clean up, it is a small enough data set.

File: drivers/hmis_external_apis/lib/tasks/import.rake (hunk 2 above)
Comment (gigxz) [reply drivers/hmis_external_apis/lib/tasks/import.rake]: added, thanks

**Pull request merged**